File: modules/shared/py_packages/dw/state_machine/mqtt_client.py
"""
MQTT Client Module

This module contains the MQTTClient class which is an observer of BaseState and handles 
MQTT communication.
"""

import logging
from abc import ABC
import json
import paho.mqtt.client as mqtt
from .base_state import BaseState

class MQTTClient(ABC):
    """
    MQTTClient Class

    This class handles MQTT communication and observes the BaseState for changes.
    """

    def __init__(self, client_id, game_state: BaseState, broker='mqtt-broker', port=1883, enable_subscriptions=None):
        """
        Initializes the MQTTClient with the given BaseState.

        Parameters:
        service_state (ServiceState): The ServiceState instance to observe.
        """
        self.client_id = client_id
        self.game_state = game_state  
        self.client = mqtt.Client(client_id=self.client_id)
        self.client.on_connect = self._on_connect
        self.client.on_disconnect = self._on_disconnect
        self.client.on_message = self._on_message
        self.connected = False

        if not enable_subscriptions:
            self.enable_subscriptions = [
                'game/state',
                'game/state_transition',
                'game/countdown',
                'game/ball_position',            
                'game/bottom_paddle_position',
                'game/top_paddle_position',
                'game/bottom_score',
                'game/top_score',
            ]
        else:
            self.enable_subscriptions = enable_subscriptions

        self.topic_to_state_map = {
            'game/state': 'game_state',
            'game/state_transition': 'game_state_transition',
            'game/countdown': 'game_countdown',
            'game/ball_position': 'game_ball_position',            
            'game/bottom_paddle_position': 'game_bottom_paddle_position',
            'game/top_paddle_position': 'game_top_paddle_position',
            'game/bottom_score': 'game_bottom_score',
            'game/top_score': 'game_top_score',
        }

        self.state_to_topic_map = {
            'game_state': 'game/state',
            'game_state_transition': 'game/state_transition',
            'game_countdown': 'game/countdown',
            'game_ball_position': 'game/ball_position',
            'game_bottom_paddle_position': 'game/bottom_paddle_position',
            'game_top_paddle_position': 'game/top_paddle_position',
            'game_bottom_score': 'game/bottom_score',
            'game_top_score': 'game/top_score',
        }

    def _on_connect(self, client, userdata, flags, rc):
        """
        Callback method for when the client connects to the broker.

        Parameters:
        client (mqtt.Client): The MQTT client instance.
        userdata (any): The private user data.
        flags (dict): Response flags sent by the broker.
        rc (int): The connection result.
        """
        self.connected = True
        for topic in self.enable_subscriptions:
            logging.info("MQTT sub: %s", topic)
            client.subscribe(topic)

    def is_connected(self):
        """
        """
        # return self.connected;
        return self.client.is_connected();


    def _on_disconnect(self, client, userdata, rc):
        """
        Callback method for when the client disconnects from the broker.

        Parameters:
        client (mqtt.Client): The MQTT client instance.
        userdata (any): The private user data.
        rc (int): The disconnection result.
        """
        self.connected = False

    def _on_message(self, client, userdata, msg):
        """
        Callback method for when a message is received from the broker.

        Parameters:
        client (mqtt.Client): The MQTT client instance.
        userdata (any): The private user data.
        msg (mqtt.MQTTMessage): The received message.
        """
        logging.info("MQTT msg")
        logging.info("MQTT msg.topic: %s, msg.payload: %s", msg.topic, msg.payload.decode())
        state_key = self.topic_to_state_map.get(msg.topic)
        logging.info("MQTT state_key: %s", state_key)
        if state_key:
            payload_str = msg.payload.decode('utf-8')
            payload_json = json.loads(payload_str)
            self.game_state.set_state(state_key, payload_json[f"{state_key}"], caller=self)
            # self.game_state.set_state(state_key, msg.payload.decode(), caller=self)

    def start(self):
        """
        Starts the MQTTClient by connecting to the broker and observing the BaseState.
        """
        self.client.connect_async("mqtt-broker", port=1883, keepalive=60)
        self.client.loop_start()
        self.game_state.add_observer(self, self._on_state_change)

    def _on_state_change(self, changed_state):
        """
        Callback method for when the BaseState changes.

        Parameters:
        changed_state (dict): The changed state values.
        """
        for key, value in changed_state.items():
            logging.info(f"key: {key}, value: {value}")
            topic = self.state_to_topic_map.get(key)
            logging.info(f"topic: {topic}")
            if topic:
                jsonValue = f"{{{topic}: {value}}}"
                json.dumps(value)
                self.client.publish(topic, value)
                # self.client.publish(topic, value)

File: modules/shared/js_packages/dw/state-machine/lib/MQTTClient.js
/**
 * @file MQTTClient.js
 * @description Implements an MQTT client using 'mqtt.js'.
 */
import mqtt from 'mqtt';
// const mqtt = require('mqtt');

// class MQTTClient {
export default class MQTTClient {
    constructor(broker_url, client_id, baseState) {
        // this.last_in_timestamp = Date.now();
        // this.last_out_timestamp = Date.now();
        this.broker_url = broker_url;
        this.client_id = client_id;
        this.baseState = baseState;
        this.connected = false;
        this.subscribeMap = {
            'game/state': 'game_state',
            'game/state_transition': 'game_state_transition',
            'game/countdown': 'game_countdown',
            'game/ball_position': 'game_ball_position',            
            'game/bottom_paddle_position': 'game_bottom_paddle_position',
            'game/top_paddle_position': 'game_top_paddle_position',
            'game/bottom_score': 'game_bottom_score',
            'game/top_score': 'game_top_score',
        };
        this.publishMap = {
            'game_state': 'game/state',
            'game_state_transition': 'game/state_transition',
            'game_countdown': 'game/countdown',
            'game_ball_position': 'game/ball_position',
            'game_bottom_paddle_position': 'game/bottom_paddle_position',
            'game_top_paddle_position': 'game/top_paddle_position',
            'game_bottom_score': 'game/bottom_score',
            'game_top_score': 'game/top_score',
        };
    }

    /**
     * Connects to the MQTT broker.
     */
    start() {
        console.log("MQTTClient broker_url: %s", this.broker_url)
        // this.client = mqtt.connect('ws://localhost:9001', {
        this.client = mqtt.connect(this.broker_url, {
            clientId: this.client_id,
            // keepalive: 60
        });

        this.client.on('connect', this.onConnect.bind(this));
        this.client.on('disconnect', this.onDisconnect.bind(this));
        this.client.on('message', this.onMessage.bind(this));
        this.baseState.addObserver(this, this.onStateChange.bind(this));        
    }

    stop() {
        this.client.end();
    }
    
    /**
     * Handles MQTT connect event.
     */
    onConnect() {
        console.log("MQTT onConnect")
        this.connected = true;
        for (let topic in this.subscribeMap) {
            console.log(`subscribe: ${topic}`);
            this.client.subscribe(topic);
        }
    }

    /**
     * Handles MQTT disconnect event.
     */
    onDisconnect() {
        console.log("MQTT onDisconnect")
        this.connected = false;
    }

    /**
     * Handles MQTT message event.
     * @param {string} topic - The topic of the message.
     * @param {Buffer} message - The message payload.
     */
    onMessage(topic, message) {
        // const timestamp = Date.now();
        // console.log(`IN - Current timestamp in milliseconds: ${timestamp - this.last_in_timestamp}`);  
        // this.last_in_timestamp = timestamp;      
        const stateKey = this.subscribeMap[topic];
        if (stateKey) {
            const theObject = JSON.parse(message.toString())
            console.log(`Message topic: ${topic}, value: ${theObject}`);
            this.baseState.setState(stateKey, theObject, this);
        }
    }

    /**
     * .
     */
    isConnected() {
        return this.connected;
    }

    /**
     * Handles state changes from BaseState.
     * @param {Object} changedState - The changed state values.
     */
    onStateChange(changedState) {
        // console.log("MQTT onStateChange")
        for (let key in changedState) {
            const topic = this.publishMap[key];
            if (topic) {
                const jsonString = JSON.stringify(changedState[key]);
                // const timestamp = Date.now();
                // console.log(`OUT - Current timestamp in milliseconds: ${timestamp - this.last_out_timestamp}`);
                // this.last_out_timestamp = timestamp;
                console.log(`Publish topic: ${topic}, value: ${jsonString}`);
                this.client.publish(topic, jsonString);
            }
        }
    }
}

// module.exports = MQTTClient;
// export default MQTTClient;